### PR TITLE
Misc cleanup

### PR DESCRIPTION
**Pull request by dmathieu**
This includes a fairly big cleanup of the packages, basically achieving the following:

* Pass golint.
* No more global flags. Provide options and pass local arguments instead.
* Remove the util package.
* Try to improve lisibility here and there.

### CODE REVIEW

**Comment by dmathieu** [comment]
@freeformz: is there anything in here which doesn't look like a go idiom, or that you would do differently?

File: broker/io.go [hunk 1]
```diff
@@ -59,14 +63,15 @@ type reader struct {
 	buffered bool
 }
 
+// NewReader creates a new io read closer
```
Comment (freeformz) [inline broker/io.go]: This documentation doesn't actually explain much.

File: broker/io_test.go [hunk 2]
```diff
@@ -209,15 +209,15 @@ func TestRedisReadFromClosed(t *testing.T) {
 	assert.Equal(t, err, io.EOF)
 
 	// We'll get true here because r.closed is already set
-	assert.True(t, ReaderDone(r))
+	assert.True(t, readerDone(r))
```
Comment (freeformz) [inline broker/io_test.go]: assert libs aren't idiomatic. I much prefer something like this:

```go
if ! readerDone(r) {
  t.Errorf("readerDone(r) != true. Should be true because <reason>")
}
```

This is just straight forward go, so there is no cognitive overhead wrt what assert.True is doing or what it will produce.

File: broker/redis.go [hunk 3]
```diff
@@ -69,26 +69,29 @@ func (c channel) id() string {
...
+func (c channel) doneID() string {
 	return string(c) + ":done"
 }
 
-func (c channel) killId() string {
+func (c channel) killID() string {
 	return string(c) + ":kill"
 }
 
+// RedisRegistrar is a channel storing data on redis
```
Comment (freeformz) [inline broker/redis.go]: If it's not already documented, we may want a doc.go that describes what a "channel" is. I keep reading that and thinking of a go channel, but that's not what it's talking about. Maybe also consider renaming channel to something else to avoid confusion.

File: logging/main.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package util
```
Comment (freeformz) [inline logging/main.go]: probably don't want to call this file "main.go". There's nothing wrong with it technically though.

File: logging/main.go [hunk 5]
```diff
@@ -10,7 +10,8 @@ import (
 const prefix = "busl"
 const defaultDomain = "development"
 
-func init() {
+// Configure configures the default logger
+func Configure() {
```
Comment (freeformz) [inline logging/main.go]: What was wrong with init?

File: cmd/busl/main.go [hunk 6]
```diff
@@ -5,21 +5,78 @@ import (
...
+	HTTPPort         = flag.String("httpPort", os.Getenv("PORT"), "HTTP port for the server.")
+	HTTPReadTimeout  = flag.Duration("httpReadTimeout", time.Hour, "Timeout for HTTP request reading")
+	HTTPWriteTimeout = flag.Duration("httpWriteTimeout", time.Hour, "Timeout for HTTP request writing")
+
+	RollbarEnvironment = flag.String("rollbarEnvironment", os.Getenv("ROLLBAR_ENVIRONMENT"), "Rollbar Enviornment for this application (development/staging/production).")
+	RollbarToken       = flag.String("rollbarToken", os.Getenv("ROLLBAR_TOKEN"), "Rollbar Token for sending issues to Rollbar.")
 )
 
 func main() {
+	logging.Configure()
```
Comment (freeformz) [inline cmd/busl/main.go]: What is it configuring and how?

File: logging/main.go (hunk 5 above)
Comment (dmathieu) [inline logging/main.go]: It's loaded very obscurely, by silently importing util. Now we're explicitely saying we want it configured and when.

File: cmd/busl/main.go [hunk 7]
```diff
@@ -5,21 +5,78 @@ import (
...
+	s := server.NewServer(conf)
+	s.ReadTimeout = *HTTPReadTimeout
+	s.WriteTimeout = *HTTPWriteTimeout
+	s.Start(*HTTPPort, awaitSignals(syscall.SIGURG))
+}
+
+func parseFlags() (*server.Config, error) {
+	conf := &server.Config{}
+
+	flag.StringVar(&conf.Credentials, "creds", os.Getenv("CREDS"), "user1:pass1|user2:pass2")
```
Comment (freeformz) [inline cmd/busl/main.go]: Why are some of these flags here and some above?

File: cmd/busl/main.go (hunk 6 above)
Comment (dmathieu) [inline cmd/busl/main.go]: It sets the log [prefix](https://golang.org/pkg/log/#SetPrefix) and [flags](https://golang.org/pkg/log/#SetFlags).
Would you have this named differently?

File: server/server.go [hunk 8]
```diff
@@ -1,166 +1,70 @@
...
-	registrar := broker.NewRedisRegistrar()
-
-	if err := registrar.Register(key(r)); err != nil {
-		http.Error(w, "Unable to create stream. Please try again.", http.StatusServiceUnavailable)
-		rollbar.Error(rollbar.ERR, fmt.Errorf("unable to register stream: %#v", err))
-		util.CountWithData("put.create.fail", 1, "error=%s", err)
-		return
+// NewServer creates a new server instance
+func NewServer(config *Config) *Server {
+	s := &Server{
```
Comment (freeformz) [inline server/server.go]: `return &Server {...}`

File: server/server.go [hunk 9]
```diff
@@ -1,166 +1,70 @@
...
-	}
-
-	netErr, ok := err.(net.Error)
-	if ok && netErr.Timeout() {
-		util.CountWithData("server.pub.read.timeout", 1, "msg=\"%v\"", err.Error())
-		handleError(w, r, netErr)
-		return
-	}
+// Start starts the server instance
+func (s *Server) Start(port string, shutdown <-chan struct{}) {
```
Comment (freeformz) [inline server/server.go]: This server isn't very composable. Why is it not a http.Handler?

Comment (dmathieu) [inline server/server.go]: Good idea. I'll refactor it away into one.

**Comment by freeformz** [comment]
I made some comments, but the PR is huge. Any chance to break this up into smaller pieces?

File: logging/main.go (hunk 5 above)
Comment (freeformz) [inline logging/main.go]: It's still pretty obscure. Your configuring it without passing anything. So how is it configured?

**Comment by dmathieu** [comment]
I've split this into commits showing my iterations process. Splitting this into several PRs without having conflicts would be difficult though.

File: assets/cat.go [hunk 10]
```diff
@@ -1,18 +0,0 @@
-package assets
-
-var HttpCatGone = `░░░░░░░░░░██░░░░░░░░░░██░░░░░░░░
```
Comment (naaman) [inline assets/cat.go]: 😢

File: server/errors.go [hunk 11]
```diff
@@ -4,20 +4,36 @@ import (
...
-	"github.com/heroku/busl/assets"
 	"github.com/heroku/busl/broker"
+	"github.com/heroku/busl/logging"
 	"github.com/heroku/busl/storage"
-	"github.com/heroku/busl/util"
 )
 
 var errNoContent = errors.New("No Content")
 
+const asciiGone = `░░░░░░░░░░██░░░░░░░░░░██░░░░░░░░
```
Comment (naaman) [inline server/errors.go]: oh hey, there it is 😸

File: broker/io_test.go (hunk 2 above)
Comment (dmathieu) [inline broker/io_test.go]: Maybe. But that'd mean refactoring all tests in the entire project. I think this PR is already big enough :)

File: server/server.go (hunk 9 above)
Comment (dmathieu) [inline server/server.go]: I'm keeping this in mind, but fear this might raise the PR's size quite a bit. Since this is how the server currently is in master, I believe we can roll with those changes, and do a proper refactoring of it afterwards.

File: broker/redis.go (hunk 3 above)
Comment (dmathieu) [inline broker/redis.go]: How about renaming "channel" to "feed"?

Comment (dmathieu) [inline broker/redis.go]: Then again, I'd suggest doing so in a follow-up PR.

**Comment by dmathieu** [comment]
Note that this PR does look long. Most of it is vendor though.

File: README.md [hunk 12]
```diff
@@ -37,7 +37,6 @@ and then:
 ```sh
 $ godep go install
 $ cp .sample.env .env
-$ export $(cat .env)
```
Comment (tt) [inline README.md]: What's the motivation for replacing this with a library that's also used in production?

File: broker/io.go [hunk 13]
```diff
@@ -6,15 +6,19 @@ import (
...
-	"github.com/heroku/busl/util"
+	"github.com/heroku/busl/logging"
 )
 
 type writer struct {
 	channel channel
 }
 
-var ErrNotRegistered = errors.New("Channel is not registered.")
+// known errors
```
Comment (tt) [inline broker/io.go]: Do we need this comment?

File: broker/io_test.go [hunk 14]
```diff
@@ -9,29 +9,29 @@ import (
...
+	uuid := uuid.NewV4().String()
 	registrar.Register(uuid)
 	r, _ := NewReader(uuid)
 	w, _ := NewWriter(uuid)
 
 	return r, w
 }
 
-func ExamplePubSub() {
+func Example_pub_sub() {
```
Comment (tt) [inline broker/io_test.go]: Is this really the preferred format?

File: broker/io_test.go (hunk 2 above)
Comment (tt) [inline broker/io_test.go]: I agree.

File: broker/redis.go (hunk 3 above)
Comment (tt) [inline broker/redis.go]: I think "channel" is well-established for the publish-subscribe pattern so I would prefer not renaming even if Go also uses that term. "Topic" would probably work too.

Comment (tt) [inline broker/redis.go]: I agree that this should be a separate pull request, though.

File: broker/io_test.go (hunk 14 above)
Comment (dmathieu) [inline broker/io_test.go]: Yes. golint complains without the underscores.
https://blog.golang.org/examples#TOC_4.

File: broker/io.go (hunk 13 above)
Comment (dmathieu) [inline broker/io.go]: Yes. golint asks for a comment for the var block.

File: README.md (hunk 12 above)
Comment (dmathieu) [inline README.md]: Not having to pollute the global dev environment with environment variables, which are very easy to forget when starting work, and won't be removed unless the terminal is closed and reopened afterwards.

File: broker/redis.go (hunk 3 above)
Comment (dmathieu) [inline broker/redis.go]: Topic looks good to me.
